core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -81,23 +85,60 @@ private[coordinator] object TransactionMetadata {
       case unknown => throw new IllegalStateException("Unknown transaction state byte " + unknown + " from the transaction status message")
     }
   }
+
+  def isValidTransition(oldState: TransactionState, newState: TransactionState): Boolean = TransactionMetadata.validPreviousStates(newState).contains(oldState)
+
+  private val validPreviousStates: Map[TransactionState, Set[TransactionState]] =
+    Map(Empty -> Set(),
+      Ongoing -> Set(Ongoing, Empty, CompleteCommit, CompleteAbort),
+      PrepareCommit -> Set(Ongoing),
+      PrepareAbort -> Set(Ongoing),
+      CompleteCommit -> Set(PrepareCommit),
+      CompleteAbort -> Set(PrepareAbort))
 }
 
 @nonthreadsafe
 private[coordinator] class TransactionMetadata(val pid: Long,
                                                var epoch: Short,
                                                val txnTimeoutMs: Int,
-                                               var state: TransactionState) {
-
-  def this(pid: Long, epoch: Short, txnTimeoutMs: Int) = this(pid, epoch, txnTimeoutMs, Empty)
+                                               var state: TransactionState,
+                                               val topicPartitions: mutable.Set[TopicPartition]) {
 
-  // participated partitions in this transaction
-  val topicPartitions = mutable.Set.empty[TopicPartition]
+  // pending state is used to indicate the state that this transaction is going to
+  // transit to, and for blocking future attempts to transit it again if it is not legal;
+  // initialized as the same as the current state
+  var pendingState: Option[TransactionState] = None
 
-  def addPartitions(partitions: Set[TopicPartition]): Unit = {
+  def addPartitions(partitions: collection.Set[TopicPartition]): Unit = {
     topicPartitions ++= partitions
   }
 
+  def prepareTransitionTo(newState: TransactionState): Boolean = {
+    if (pendingState.isDefined)
+      throw new IllegalStateException(s"Preparing transaction state transition to $newState while it already a pending state ${pendingState.get}")
```

Comment:
We always check `isValidTransition` before calling this function, so it should be a `should-never-happen` error.